### PR TITLE
get_substituted_config bugfix; variable substitution

### PR DESCRIPTION
**Pull request by roblim**
- added additional test case in test_checkpoint for nested checkpoint config with different template name at runtime
- fixed bug in `Checkpoint.get_substituted_config` that occurred when checkpoint already had substituted config and a new template name was passed in at runtime (issue happened during subsequent recursive call)
- added variable substitution for checkpoint config - it's almost the same as the processing in DataContext, might want to factor out into a class method or similar

### CODE REVIEW

File: great_expectations/checkpoint/checkpoint.py
Comment (alexsherstinsky) [inline great_expectations/checkpoint/checkpoint.py]: Minor: If this is a constant, shouldn't it be `BaseDataContext. DOLLAR_SIGN_ESCAPE_STRING`?

File: great_expectations/checkpoint/checkpoint.py
Comment (alexsherstinsky) [inline great_expectations/checkpoint/checkpoint.py]: Note: The code in `_get_substituted_validation_dict` changed -- I do not know why github is not telling us to pull the latest from `checkpoint-class`.

File: great_expectations/core/util.py
Comment (alexsherstinsky) [inline great_expectations/core/util.py]: Opportunity: We now have `parse_string_to_datetime()` in `great_expectations/datasource/data_connector/sorter/date_time_sorter.py` and `get_datetime_string_from_strftime_format()` in `great_expectations/core/util.py` -- it would be nice to combine -- if appropriate -- and reduce code duplication...

File: great_expectations/data_context/data_context.py
Comment (alexsherstinsky) [inline great_expectations/data_context/data_context.py]: Same comment as in checkpoint.py -- we should probably use `BaseDataContext. DOLLAR_SIGN_ESCAPE_STRING`.

File: tests/checkpoint/test_checkpoint.py
Comment (alexsherstinsky) [inline tests/checkpoint/test_checkpoint.py]: I am not sure whether or not this fixture is needed.  I added a bunch of tests, including those for the nested Checkpoint YML configuration (above this fixture).  Maybe all we need to do is uncommented the variable substitution, remove comments containing "ALEX", and make sure that those tests pass?  Just wondering.  Thanks.

Comment (alexsherstinsky) [reply tests/checkpoint/test_checkpoint.py]: If we decide that this fixture and tests should still be here, then perhaps we should either put the fixture at the top or move it to conftest.py (I did the latter for my fixture).

File: tests/checkpoint/test_checkpoint.py
Comment (alexsherstinsky) [inline tests/checkpoint/test_checkpoint.py]: Why are we commenting out the line with `$VAR`?  I thought that this PR would make it work?..

File: tests/checkpoint/test_checkpoint.py
Comment (alexsherstinsky) [inline tests/checkpoint/test_checkpoint.py]: Same comment as above -- I thought that `$MY_PARAM` would work now thanks to this PR?..

File: tests/checkpoint/test_checkpoint.py
Comment (alexsherstinsky) [inline tests/checkpoint/test_checkpoint.py]: Same comment as above -- I thought that `$MY_PARAM` would work now...

File: tests/checkpoint/test_checkpoint.py
Comment (alexsherstinsky) [inline tests/checkpoint/test_checkpoint.py]: Same comment as above -- could we reuse the tests I created by simply uncommenting out the `$MY_VAR` type statements and removing comments with my name in them?..

File: tests/core/test_util.py
Comment (alexsherstinsky) [inline tests/core/test_util.py]: So cool!

**Review by alexsherstinsky — COMMENTED** [review]
Will be happy to approve once the answers to my questions are provided, this branch is synchronized with `checkpoint-class` upstream, and all tests pass.  Please let us make sure that we merge `https://github.com/great-expectations/great_expectations/pull/2228` before merging this PR (2227); otherwise, we will be faced with difficult merge conflicts.

Importantly the existing tests I was referring to (where we would uncomment the dollar sign evaluation parameters and remove comments with my name in them) will be available once we merge the above PR (2228).  Thank you very much.

File: great_expectations/checkpoint/checkpoint.py
Comment (roblim) [reply great_expectations/checkpoint/checkpoint.py]: Good point - thanks for always helping me make my code better! I suppose I used `self.data_context` since in DataContext, `self.DOLLAR_SIGN_ESCAPE_STRING` is used, but now I see that that makes sense because same inheritance chain. Will change.

File: great_expectations/checkpoint/checkpoint.py
Comment (roblim) [reply great_expectations/checkpoint/checkpoint.py]: Oh. Do you remember the way it changed that affects the run_time defined here?

File: great_expectations/core/util.py
Comment (roblim) [reply great_expectations/core/util.py]: I think these are different though, since in `parse_string_to_datetime`, it takes in a `str` that's semantically a date and returns a `datetime.date` object, and in this function, it takes in a strftime format `str` like `%m%A` and optionally a datetime object, and returns a semantic date `str`. But perhaps these should be together rather than in separate places?

File: great_expectations/data_context/data_context.py
Comment (roblim) [reply great_expectations/data_context/data_context.py]: Agreed - also adjusted.

File: tests/checkpoint/test_checkpoint.py
Comment (roblim) [reply tests/checkpoint/test_checkpoint.py]: Okay - makes total sense, since I borrowed a lot from a test you created before, before there was an actual fixture. I think I maybe have added a few things, but if their functionally the same, I agree we should just have one and put it in conftest. Will examine.

File: tests/checkpoint/test_checkpoint.py
Comment (roblim) [reply tests/checkpoint/test_checkpoint.py]: Oh. Not commenting out or removing anything here. I agree it's a bit confusing since there's tons of stuff in those tests.

The above change is defined within the expected output, hence all the replacements since var substitution works now.

File: tests/checkpoint/test_checkpoint.py
Comment (roblim) [reply tests/checkpoint/test_checkpoint.py]: See above comment (this is in the "expected" part)

File: tests/checkpoint/test_checkpoint.py
Comment (roblim) [reply tests/checkpoint/test_checkpoint.py]: See above comment (this is in the "expected" part)

File: tests/checkpoint/test_checkpoint.py
Comment (roblim) [reply tests/checkpoint/test_checkpoint.py]: Probably - agree that we should avoid duplication.

File: tests/core/test_util.py
Comment (roblim) [reply tests/core/test_util.py]: Thx! Though I think I misunderstood James ask with the eval param processing. Will clear this up.

File: great_expectations/core/util.py
Comment (alexsherstinsky) [reply great_expectations/core/util.py]: @roblim I am guessing that "together" might be nicer...  Your call as to when/how.

File: tests/checkpoint/test_checkpoint.py
Comment (roblim) [reply tests/checkpoint/test_checkpoint.py]: - After going over the tests and fixtures, I think we should keep the ones I added since they also test running checkpoints with runtime kwargs, as well as more deeply nested configs
- I moved the fixture I added with all the checkpoint templates into conftest and renamed it to be more clear

Per conversation I had with @jcampbell, I'm holding off on eval param processing for now, since this is a new feature that deserves its own design review. Therefore, I've left your "Alex" comments relating to the eval params keys. I did however uncomment the parts that were purely var sub related (run_name key) and updated the test, which is passing.

**Comment by roblim** [comment]
@alexsherstinsky - I think I've addressed all your comments. Please re-review - would be great to merge. Thanks!

**Review by alexsherstinsky — APPROVED** [review]
Please remove comments that have the string Alex/Rob as appropriate (but please leave the other ALEX comments intact).

**Pull request closed without merging**
